File: ArchLucid.Decisioning/Findings/DecisionGradeFindingProvenanceValidator.cs
using ArchLucid.Contracts.Findings;
using ArchLucid.Contracts.Findings.Payloads;
using ArchLucid.Core.Findings;

namespace ArchLucid.Decisioning.Findings;

/// <summary>TB-2227: structural provenance required for decision-grade findings at commit.</summary>
public static class DecisionGradeFindingProvenanceValidator
{
    public static IReadOnlyList<string> GetViolations(FindingsSnapshot snapshot)
    {
        ArgumentNullException.ThrowIfNull(snapshot);

        List<string> violations = [];

        foreach (Finding finding in snapshot.Findings)
        {
            if (finding.Classification == FindingClassification.ChecklistCoverage)
                continue;

            string? violation = GetViolation(finding);

            if (violation is not null)
                violations.Add(violation);
        }

        return violations;
    }

    private static string? GetViolation(Finding finding)
    {
        if (InsightDensityFindingSourceClassifier.IsAgentArchitectureFinding(finding.FindingType))
            return HasAgentCitationProvenance(finding)
                ? null
                : $"Finding '{finding.FindingId}' ({finding.FindingType}) lacks agent citation provenance.";

        return HasTypedEngineProvenance(finding)
            ? null
            : $"Finding '{finding.FindingId}' ({finding.FindingType}) lacks typed-engine provenance.";
    }

    private static bool HasAgentCitationProvenance(Finding finding)
    {
        if (finding.Trace?.Citations is { Count: > 0 } citations
            && citations.Any(static c => !string.IsNullOrWhiteSpace(c)))
            return true;

        return false;
    }

    private static bool HasTypedEngineProvenance(Finding finding)
    {
        if (HasInventoryDrivenEngineProvenance(finding))
            return true;

        if (HasGraphCoverageEngineProvenance(finding))
            return true;

        if (HasWorkloadExpectationEngineProvenance(finding))
            return true;

        if (HasCrossRunDiffEngineProvenance(finding))
            return true;

        bool hasNodes = finding.RelatedNodeIds is { Count: > 0 } nodes
                        && nodes.Any(static n => !string.IsNullOrWhiteSpace(n));

        bool hasRules = finding.Trace?.RulesApplied is { Count: > 0 } rules
                        && rules.Any(static r => !string.IsNullOrWhiteSpace(r));

        if (hasNodes && hasRules)
            return true;

        if (finding.Trace?.Citations is { Count: > 0 } citations
            && citations.Any(static c => !string.IsNullOrWhiteSpace(c)))
            return true;

        return false;
    }

    private static bool HasInventoryDrivenEngineProvenance(Finding finding)
    {
        if (!IsInventoryDrivenEngine(finding.EngineType))
            return false;

        bool hasRules = finding.Trace?.RulesApplied is { Count: > 0 } rules
                        && rules.Any(static r => !string.IsNullOrWhiteSpace(r));

        if (!hasRules)
            return false;

        if (finding.RelatedNodeIds is { Count: > 0 } nodes
            && nodes.Any(static n => !string.IsNullOrWhiteSpace(n)))
        {
            return true;
        }

        if (finding.Payload is InventoryReconciliationFindingPayload reconciliation)
        {
            return reconciliation.GraphOnlyResourceIds.Any(static id => !string.IsNullOrWhiteSpace(id))
                   || reconciliation.InventoryOnlyResourceIds.Any(static id => !string.IsNullOrWhiteSpace(id));
        }

        if (finding.Payload is RequirementFindingPayload requirement)
            return !string.IsNullOrWhiteSpace(requirement.RequirementName);

        if (finding.Payload is AdvisorCostRecommendationFindingPayload advisorCost)
            return !string.IsNullOrWhiteSpace(advisorCost.RecommendationId);

        return false;
    }

    private static bool HasGraphCoverageEngineProvenance(Finding finding)
    {
        if (!IsGraphCoverageEngine(finding.EngineType))
            return false;

        bool hasRules = finding.Trace?.RulesApplied is { Count: > 0 } rules
                        && rules.Any(static r => !string.IsNullOrWhiteSpace(r));

        if (!hasRules)
            return false;

        if (finding.RelatedNodeIds is { Count: > 0 } nodes
            && nodes.Any(static n => !string.IsNullOrWhiteSpace(n)))
        {
            return true;
        }

        if (finding.Payload is PolicyCoverageFindingPayload)
            return true;

        if (finding.Payload is TopologyCoverageFindingPayload)
            return true;

        if (finding.Payload is RequirementCoverageFindingPayload requirement)
        {
            return requirement.UncoveredRequirementCount > 0
                   || requirement.UncoveredRequirements.Any(static id => !string.IsNullOrWhiteSpace(id));
        }

        if (finding.Payload is SecurityCoverageFindingPayload security)
        {
            return security.UnprotectedResourceCount > 0
                   || security.UnprotectedResources.Any(static id => !string.IsNullOrWhiteSpace(id));
        }

        return false;
    }

    private static bool HasWorkloadExpectationEngineProvenance(Finding finding)
    {
        if (!IsWorkloadExpectationEngine(finding.EngineType))
            return false;

        bool hasRules = finding.Trace?.RulesApplied is { Count: > 0 } rules
                        && rules.Any(static r => !string.IsNullOrWhiteSpace(r));

        if (!hasRules)
            return false;

        if (finding.RelatedNodeIds is { Count: > 0 } nodes
            && nodes.Any(static n => !string.IsNullOrWhiteSpace(n)))
        {
            return true;
        }

        if (finding.Payload is RequirementExpectationFindingPayload requirement)
        {
            return requirement.TopologyNodeCount > 0
                   || requirement.MissingThemes.Any(static theme => !string.IsNullOrWhiteSpace(theme));
        }

        if (finding.Payload is SecurityBaselineExpectationFindingPayload security)
        {
            return security.TopologyNodeCount > 0
                   || security.MissingCategories.Any(static category => !string.IsNullOrWhiteSpace(category));
        }

        if (finding.Payload is SecurityBaselineCompletenessFindingPayload completeness)
        {
            return completeness.TopologyNodeCount > 0
                   || completeness.MissingControlFamilies.Any(static family => !string.IsNullOrWhiteSpace(family));
        }

        return false;
    }

    private static bool IsWorkloadExpectationEngine(string? engineType) =>
        string.Equals(engineType, "requirement-expectation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "security-baseline-expectation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "security-baseline-completeness", StringComparison.OrdinalIgnoreCase);

    private static bool HasCrossRunDiffEngineProvenance(Finding finding)
    {
        if (!IsCrossRunDiffEngine(finding.EngineType))
            return false;

        bool hasRules = finding.Trace?.RulesApplied is { Count: > 0 } rules
                        && rules.Any(static r => !string.IsNullOrWhiteSpace(r));

        if (!hasRules)
            return false;

        if (finding.RelatedNodeIds is { Count: > 0 } nodes
            && nodes.Any(static n => !string.IsNullOrWhiteSpace(n)))
        {
            return true;
        }

        if (finding.Payload is TopologyGapFindingPayload gap)
            return !string.IsNullOrWhiteSpace(gap.GapCode);

        if (finding.Payload is RequirementCoverageFindingPayload requirement)
        {
            return requirement.UncoveredRequirementCount > 0
                   || requirement.UncoveredRequirements.Any(static id => !string.IsNullOrWhiteSpace(id));
        }

        if (finding.Payload is TopologyCoverageFindingPayload topology)
        {
            return topology.TopologyNodeCount > 0
                   || topology.PresentCategories.Any(static category => !string.IsNullOrWhiteSpace(category))
                   || topology.ExpectedCategories.Any(static category => !string.IsNullOrWhiteSpace(category));
        }

        return false;
    }

    private static bool IsCrossRunDiffEngine(string? engineType) =>
        string.Equals(engineType, "requirement-cross-run-diff", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "topology-cross-run-diff", StringComparison.OrdinalIgnoreCase);

    private static bool IsGraphCoverageEngine(string? engineType) =>
        string.Equals(engineType, "policy-coverage", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "topology-coverage", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "requirement-coverage", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "security-coverage", StringComparison.OrdinalIgnoreCase);

    private static bool IsInventoryDrivenEngine(string? engineType) =>
        string.Equals(engineType, "azure-inventory-reconciliation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "aws-inventory-reconciliation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "gcp-inventory-reconciliation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "azure-inventory-security-baseline", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "aws-inventory-security-baseline", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "gcp-inventory-security-baseline", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "orphaned-azure-resource", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "orphaned-aws-resource", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "orphaned-gcp-resource", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "advisor-cost-recommendation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "aws-cost-recommendation", StringComparison.OrdinalIgnoreCase)
        || string.Equals(engineType, "gcp-cost-recommendation", StringComparison.OrdinalIgnoreCase);
}

File: archlucid-ui/src/lib/findings/decision-grade-finding-provenance-validator.ts
import { FINDING_CLASSIFICATION_CHECKLIST_COVERAGE } from "@/lib/findings/review-detail-findings-classification-band";
import { normalizeFindingTrustLabelName } from "@/lib/findings/finding-provenance-display";

type ParsedFinding = {
  readonly findingId: string;
  readonly findingType: string;
  readonly classification: string | null;
  readonly engineType: string | null;
  readonly relatedNodeIds: readonly string[];
  readonly traceCitations: readonly string[];
  readonly traceRulesApplied: readonly string[];
  readonly payload: Record<string, unknown> | null;
};

function readTrimmedString(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }

  const trimmed = value.trim();

  return trimmed.length > 0 ? trimmed : null;
}

function readStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value
    .filter((entry): entry is string => typeof entry === "string" && entry.trim().length > 0)
    .map((entry) => entry.trim());
}

function readClassification(value: unknown): string | null {
  const raw = readTrimmedString(value);

  if (raw === FINDING_CLASSIFICATION_CHECKLIST_COVERAGE || raw === "DecisionGradeFinding") {
    return raw;
  }

  return null;
}

function parseFinding(raw: unknown): ParsedFinding | null {
  if (raw === null || typeof raw !== "object") {
    return null;
  }

  const record = raw as Record<string, unknown>;
  const findingId = readTrimmedString(record.findingId);

  if (findingId === null) {
    return null;
  }

  const trace =
    record.trace !== null && typeof record.trace === "object"
      ? (record.trace as Record<string, unknown>)
      : null;
  const payload =
    record.payload !== null && typeof record.payload === "object"
      ? (record.payload as Record<string, unknown>)
      : null;

  return {
    findingId,
    findingType: readTrimmedString(record.findingType) ?? "",
    classification: readClassification(record.classification),
    engineType: readTrimmedString(record.engineType),
    relatedNodeIds: readStringArray(record.relatedNodeIds),
    traceCitations: trace !== null ? readStringArray(trace.citations) : [],
    traceRulesApplied: trace !== null ? readStringArray(trace.rulesApplied) : [],
    payload,
  };
}

function readFindingsFromSnapshot(findingsSnapshot: unknown): ParsedFinding[] {
  if (findingsSnapshot === null || typeof findingsSnapshot !== "object") {
    return [];
  }

  const findings = (findingsSnapshot as { findings?: unknown }).findings;

  if (!Array.isArray(findings)) {
    return [];
  }

  return findings
    .map((entry) => parseFinding(entry))
    .filter((entry): entry is ParsedFinding => entry !== null);
}

function isAgentArchitectureFinding(findingType: string): boolean {
  return findingType.startsWith("AgentArchitectureFinding");
}

const INVENTORY_DRIVEN_ENGINE_TYPES = new Set([
  "azure-inventory-reconciliation",
  "aws-inventory-reconciliation",
  "gcp-inventory-reconciliation",
  "azure-inventory-security-baseline",
  "aws-cost-recommendation",
  "gcp-cost-recommendation",
]);

function hasInventoryDrivenEngineProvenance(finding: ParsedFinding): boolean {
  const engineType = finding.engineType?.trim().toLowerCase() ?? "";

  if (!INVENTORY_DRIVEN_ENGINE_TYPES.has(engineType)) {
    return false;
  }

  if (finding.traceRulesApplied.length === 0) {
    return false;
  }

  if (finding.relatedNodeIds.length > 0) {
    return true;
  }

  const payload = finding.payload;

  if (payload === null) {
    return false;
  }

  const graphOnly = readStringArray(payload.graphOnlyResourceIds);
  const inventoryOnly = readStringArray(payload.inventoryOnlyResourceIds);

  if (graphOnly.length > 0 || inventoryOnly.length > 0) {
    return true;
  }

  const requirementName = readTrimmedString(payload.requirementName);
  const recommendationId = readTrimmedString(payload.recommendationId);

  return requirementName !== null || recommendationId !== null;
}

function readTrustLabelName(finding: ParsedFinding): string | null {
  const fromPayload = finding.payload !== null ? readTrimmedString(finding.payload.trustLabel) : null;

  return normalizeFindingTrustLabelName(fromPayload);
}

function hasResolvableSourcePointer(finding: ParsedFinding): boolean {
  if (finding.traceCitations.length > 0) {
    return true;
  }

  if (finding.payload === null) {
    return false;
  }

  const evidenceRefs = finding.payload.evidence;

  if (!Array.isArray(evidenceRefs)) {
    return false;
  }

  return evidenceRefs.some((entry) => {
    if (entry === null || typeof entry !== "object") {
      return false;
    }

    const record = entry as Record<string, unknown>;
    const artifactId = readTrimmedString(record.artifactId);
    const excerpt = readTrimmedString(record.excerpt);
    const lineRange = readTrimmedString(record.lineRange);

    return artifactId !== null || excerpt !== null || lineRange !== null;
  });
}

function hasTypedEngineProvenance(finding: ParsedFinding): boolean {
  if (hasInventoryDrivenEngineProvenance(finding)) {
    return true;
  }

  const hasNodes = finding.relatedNodeIds.length > 0;
  const hasRules = finding.traceRulesApplied.length > 0;

  if (hasNodes && hasRules) {
    return true;
  }

  return finding.traceCitations.length > 0;
}

function getViolation(finding: ParsedFinding): string | null {
  if (finding.classification === FINDING_CLASSIFICATION_CHECKLIST_COVERAGE) {
    return null;
  }

  if (isAgentArchitectureFinding(finding.findingType)) {
    if (finding.traceCitations.length > 0) {
      return null;
    }

    return `Finding '${finding.findingId}' (${finding.findingType}) lacks agent citation provenance.`;
  }

  if (hasTypedEngineProvenance(finding)) {
    if (readTrustLabelName(finding) === "EvidenceBacked" && !hasResolvableSourcePointer(finding)) {
      return `Finding '${finding.findingId}' (${finding.findingType}) is evidence-backed but lacks a resolvable source pointer.`;
    }

    return null;
  }

  return `Finding '${finding.findingId}' (${finding.findingType}) lacks typed-engine provenance.`;
}

/** Mirrors C# DecisionGradeFindingProvenanceValidator for Working career export gates (FC-38). */
export function getDecisionGradeFindingProvenanceViolations(findingsSnapshot: unknown): readonly string[] {
  const violations: string[] = [];

  for (const finding of readFindingsFromSnapshot(findingsSnapshot)) {
    const violation = getViolation(finding);

    if (violation !== null) {
      violations.push(violation);
    }
  }

  return violations;
}
